From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Tue, 13 Oct 2026 01:52:36 -0700
Subject: [PATCH] mshtml: anchor.pathname slash for about: and file:,
 normalized MediaQueryList.media

- IHTMLAnchorElement::get_pathname prepends the slash IE shows for the about:
  and file: schemes ("about:blank" -> "/blank", "file:path" -> "/path"); other
  schemes keep the pathname as the engine gives it.
- MediaQueryList.media is reported in IE's normalized form: a query starting
  with a feature gets the "all and" prefix, the space after the colon of a
  feature goes, and a media type IE does not know is called "unknown".

Removes the matching todo_wine from tests/dom.js and es5.js.

diff --git a/dlls/mshtml/htmlanchor.c b/dlls/mshtml/htmlanchor.c
index f2a085a..9d83c9a 100644
--- a/dlls/mshtml/htmlanchor.c
+++ b/dlls/mshtml/htmlanchor.c
@@ -595,13 +595,40 @@ static HRESULT WINAPI HTMLAnchorElement_get_pathname(IHTMLAnchorElement *iface,
     HTMLAnchorElement *This = impl_from_IHTMLAnchorElement(iface);
     nsAString pathname_str;
     nsresult nsres;
+    HRESULT hres;
+    IUri *uri;
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    /* FIXME: IE prepends a slash for some protocols */
     nsAString_Init(&pathname_str, NULL);
     nsres = nsIDOMHTMLAnchorElement_GetPathname(This->nsanchor, &pathname_str);
-    return return_nsstr(nsres, &pathname_str, p);
+    hres = return_nsstr(nsres, &pathname_str, p);
+    if(FAILED(hres) || !*p || **p == '/')
+        return hres;
+
+    /* IE prepends a slash for the about: and file: schemes */
+#ifdef SG_MUTANT_MSHTML_NOSLASH
+    return S_OK;
+#endif
+    uri = get_anchor_uri(This);
+    if(uri) {
+        BSTR scheme;
+
+        if(SUCCEEDED(IUri_GetSchemeName(uri, &scheme))) {
+            if(!wcsicmp(scheme, L"about") || !wcsicmp(scheme, L"file")) {
+                BSTR path = SysAllocStringLen(NULL, SysStringLen(*p) + 1);
+                if(path) {
+                    path[0] = '/';
+                    wcscpy(path + 1, *p);
+                    SysFreeString(*p);
+                    *p = path;
+                }
+            }
+            SysFreeString(scheme);
+        }
+        IUri_Release(uri);
+    }
+    return S_OK;
 }
 
 static HRESULT WINAPI HTMLAnchorElement_put_port(IHTMLAnchorElement *iface, BSTR v)
diff --git a/dlls/mshtml/omnavigator.c b/dlls/mshtml/omnavigator.c
index b954844..7fda0a5 100644
--- a/dlls/mshtml/omnavigator.c
+++ b/dlls/mshtml/omnavigator.c
@@ -2321,15 +2321,78 @@ static inline struct media_query_list *impl_from_IWineMSHTMLMediaQueryList(IWine
 DISPEX_IDISPATCH_IMPL(media_query_list, IWineMSHTMLMediaQueryList,
                       impl_from_IWineMSHTMLMediaQueryList(iface)->dispex)
 
+/* IE reports a media query list in its own normalized form: a query starting with a
+ * feature gets the "all and" prefix, spaces after the colon of a feature are removed
+ * and a media type it does not know is called "unknown". */
+static BSTR normalize_media_query(const WCHAR *query)
+{
+    static const WCHAR * const known[] = { L"all", L"aural", L"braille", L"embossed", L"handheld", L"print",
+                                           L"projection", L"screen", L"speech", L"tty", L"tv", L"not", L"only" };
+    const WCHAR *p = query ? query : L"";
+    WCHAR *buf, *out;
+    BOOL in_parens = FALSE;
+    size_t len, word_len;
+    unsigned i;
+    BSTR ret;
+
+    buf = malloc((wcslen(p) + 16) * sizeof(WCHAR));
+    if(!buf)
+        return NULL;
+    out = buf;
+
+    if(*p == '(') {
+        wcscpy(out, L"all and ");
+        out += 8;
+    }else {
+        /* leading media type (or "not" / "only") */
+        for(word_len = 0; iswalpha(p[word_len]) || p[word_len] == '-'; word_len++);
+        if(word_len) {
+            for(i = 0; i < ARRAY_SIZE(known); i++)
+                if(wcslen(known[i]) == word_len && !wcsnicmp(p, known[i], word_len))
+                    break;
+            if(i == ARRAY_SIZE(known)) {
+                wcscpy(out, L"unknown");
+                out += 7;
+                p += word_len;
+            }
+        }
+    }
+
+    for(; *p; p++) {
+        if(*p == '(') in_parens = TRUE;
+        else if(*p == ')') in_parens = FALSE;
+        *out++ = *p;
+        if(in_parens && *p == ':')
+            while(p[1] == ' ') p++;
+    }
+    *out = 0;
+    len = out - buf;
+    ret = SysAllocStringLen(buf, len);
+    free(buf);
+    return ret;
+}
+
 static HRESULT WINAPI media_query_list_get_media(IWineMSHTMLMediaQueryList *iface, BSTR *p)
 {
     struct media_query_list *media_query_list = impl_from_IWineMSHTMLMediaQueryList(iface);
     nsAString nsstr;
+    HRESULT hres;
+    BSTR media;
 
     TRACE("(%p)->(%p)\n", media_query_list, p);
 
     nsAString_InitDepend(&nsstr, NULL);
-    return return_nsstr(nsIDOMMediaQueryList_GetMedia(media_query_list->nsquerylist, &nsstr), &nsstr, p);
+    hres = return_nsstr(nsIDOMMediaQueryList_GetMedia(media_query_list->nsquerylist, &nsstr), &nsstr, &media);
+    if(FAILED(hres))
+        return hres;
+
+#ifdef SG_MUTANT_MSHTML_NOMEDIA
+    *p = media;
+    return S_OK;
+#endif
+    *p = normalize_media_query(media);
+    SysFreeString(media);
+    return *p ? S_OK : E_OUTOFMEMORY;
 }
 
 static HRESULT WINAPI media_query_list_get_matches(IWineMSHTMLMediaQueryList *iface, VARIANT_BOOL *p)
diff --git a/dlls/mshtml/tests/dom.js b/dlls/mshtml/tests/dom.js
index 256cab6..1bf12d1 100644
--- a/dlls/mshtml/tests/dom.js
+++ b/dlls/mshtml/tests/dom.js
@@ -163,7 +163,7 @@ sync_test("anchor", function() {
         { href: "https://www.winehq.org:123/about",
           protocol: "https:", host: "www.winehq.org:123", path: "/about" },
         { href: "about:blank",
-          protocol: "about:", host: "", path: "/blank", todo_pathname: 1 },
+          protocol: "about:", host: "", path: "/blank" },
         { href: "unknown:path",
           protocol: "unknown:", host: "", path: "path" },
         { href: "ftp:path",
@@ -173,7 +173,7 @@ sync_test("anchor", function() {
         { href: "ws:path",
           protocol: "ws:", host: "", path: "path" },
         { href: "file:path",
-          protocol: "file:", host: "", path: "/path", todo_pathname: 1 },
+          protocol: "file:", host: "", path: "/path" },
         { href: "file:///c:/dir/file.html",
           protocol: "file:", host: "", path: "/c:/dir/file.html" },
         { href: "http://www.winehq.org/about",
@@ -188,7 +188,6 @@ sync_test("anchor", function() {
         var anchor = document.body.firstChild;
         ok(anchor.protocol === t.protocol, "anchor(" + t.href + ").protocol = " + anchor.protocol);
         ok(anchor.host === t.host, "anchor(" + t.href + ").host = " + anchor.host);
-        todo_wine_if("todo_pathname" in t).
         ok(anchor.pathname === t.path, "anchor(" + t.href + ").pathname = " + anchor.pathname);
     }
 });
diff --git a/dlls/mshtml/tests/es5.js b/dlls/mshtml/tests/es5.js
index 9176908..634cc51 100644
--- a/dlls/mshtml/tests/es5.js
+++ b/dlls/mshtml/tests/es5.js
@@ -2673,7 +2673,6 @@ async_test("matchMedia", function() {
     ];
     for(i = 0; i < r.length; i++) {
         mql = window.matchMedia(r[i][0]);
-        todo_wine_if(r[i][0] !== 42).
         ok(mql.media === r[i][1], r[i][0] + " media = " + mql.media);
         ok(mql.matches === false, r[i][0] + " matches");
     }
